exonum/src/node/state.rs

```diff
@@ -870,9 +870,53 @@ impl State {
             }
         }
 
+        // Depending on the build type and amount of proposes, we may want
+        // to reorder proposes. See comments in both implementations of
+        // `reorder_proposes_if_needed` to get details.
+        Self::reorder_proposes_if_needed(&mut full_proposes);
+
         full_proposes
     }
 
+    #[cfg(debug_assertions)]
+    fn reorder_proposes_if_needed(full_proposes: &mut Vec<(Hash, Round)>) {
+        // For tests we want the order of proposes to be predictable,
+        // so (unlike the release version) we *always* sort by both round and hash.
+        full_proposes.sort_unstable_by(|a, b| {
+            // Compare rounds first.
+            // Note that we call `cmp` on `b` to obtain descending order.
+            let cmp_result = b.1.cmp(&a.1);
+            if let std::cmp::Ordering::Equal = cmp_result {
+                // Rounds are equal, compare by hash.
+                a.0.cmp(&b.0)
+            } else {
+                // Rounds are different, use the comparison result.
+                cmp_result
+            }
+        });
+    }
+
+    #[cfg(not(debug_assertions))]
+    fn reorder_proposes_if_needed(full_proposes: &mut Vec<(Hash, Round)>) {
+        // If we have a lot of proposes, we should sort it by the round in
+        // descending order, since it's more likely to commit a propose with
+        // greater round, and we don't want to process a lot of proposes from
+        // lower rounds.
+        // However, if the amount of proposes isn't big, overhead from sorting
+        // is not desired.
+
+        // TODO: Clarify the value for this constant (ECR-4050).
+        const MIN_PROPOSES_AMOUNT_FOR_SORTING: usize = 10;
+
+        if full_proposes.len() >= MIN_PROPOSES_AMOUNT_FOR_SORTING {
+            full_proposes.sort_unstable_by(|a, b| {
```

Comment:
Nit: Does tuple destructuring here makes the formatter go crazy? Otherwise, it could make things clearer.